### PR TITLE
Bump datafusion -> 21.0.0 and add dyn and dyn_hash functions to custo…

### PR DESCRIPTION
**Pull request by jdye64**
…m parsers

### CODE REVIEW

File: tests/integration/test_compatibility.py [hunk 1]
```diff
@@ -739,6 +739,24 @@ def test_window_sum_avg_partition_by():
...
+                    ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS a4,
+                SUM(b+a) OVER (PARTITION BY b ORDER BY a NULLS FIRST
+                    ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS a6
+            FROM a
+            ORDER BY a NULLS FIRST, b NULLS FIRST, c NULLS FIRST
+        """
+    )
+    print(f"Result Received: {result}")
+
+
```
Comment (charlesbluca) [inline tests/integration/test_compatibility.py:759]: Looks like this was only here for dev purposes?

```suggestion
```

File: tests/integration/test_create.py [hunk 2]
```diff
@@ -92,6 +92,7 @@ def test_create_from_csv_persist(c, df, temporary_data_file, gpu):
 
 def test_wrong_create(c):
     with pytest.raises(AttributeError):
+        print("Invoking c.sql()")
```
Comment (charlesbluca) [inline tests/integration/test_create.py:95]: Same here:

```suggestion
```

**Review by charlesbluca — COMMENTED** [review]
Thanks @jdye64 for opening and @ayushdg for resolving the remaining failures here 🎉 a couple comments but overall things look good here to me

File: docs/environment.yml [hunk 3]
```diff
@@ -19,4 +19,4 @@ dependencies:
   - tabulate
   - setuptools-rust>=1.5.2
   - ucx-proc=*=cpu
-    #  - rust>=1.60.0
+  - rust>=1.67.0
```
Comment (charlesbluca) [inline docs/environment.yml:22]: Might be worth adding this constraint to the rust build dependency in our nightly recipe

File: tests/integration/test_compatibility.py (hunk 1 above)
Comment (jdye64) [reply tests/integration/test_compatibility.py:759]: I will remove this in next commit

File: tests/integration/test_create.py (hunk 2 above)
Comment (jdye64) [reply tests/integration/test_create.py:95]: Will remove this in next commit.

File: continuous_integration/recipe/meta.yaml [hunk 4]
```diff
@@ -22,7 +22,7 @@ build:
 
 requirements:
   build:
-    - {{ compiler('rust') }}
+    - {{ compiler('rust') }} >=1.67.0
```
Comment (charlesbluca) [inline continuous_integration/recipe/meta.yaml:25]:
```suggestion
    - {{ compiler('rust') }} >=1.65
```

File: docs/environment.yml (hunk 3 above)
Comment (charlesbluca) [inline docs/environment.yml:22]:
```suggestion
  - rust>=1.65
```

**Review by charlesbluca — COMMENTED** [review]
Minor compiler constraint changes to resolve conda builds

File: dask_sql/physical/rel/logical/window.py [hunk 5]
```diff
@@ -246,7 +246,9 @@ def convert(self, rel: "LogicalPlan", context: "dask_sql.Context") -> DataContai
         (dc,) = self.assert_inputs(rel, 1, context)
 
         # Output to the right field names right away
-        field_names = rel.getRowType().getFieldNames()
+        field_names = rel.getRowType()
+
+        field_names = field_names.getFieldNames()
```
Comment (ayushdg) [inline dask_sql/physical/rel/logical/window.py:251]: We might be able to revert back to the old version

File: docs/environment.yml [hunk 6]
```diff
@@ -19,4 +19,4 @@ dependencies:
   - tabulate
   - setuptools-rust>=1.5.2
   - ucx-proc=*=cpu
-    #  - rust>=1.60.0
```
Comment (ayushdg) [inline docs/environment.yml:22]: Should be good to remove the line here

**Review by ayushdg — APPROVED** [review]
Minor suggestions but nothing blocking

**Pull request merged**